tensorflow_federated/python/research/robust_aggregation/rfa.py

```diff
@@ -0,0 +1,102 @@
+# Lint as: python3
+# Copyright 2019, Krishna Pillutla and Sham M. Kakade and Zaid Harchaoui.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Simple implementation of the RFA Algorithm for robust aggregation."""
+
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import six
+from six.moves import range
+
+import tensorflow as tf
+import tensorflow_federated as tff
+
+
+def build_stateless_robust_aggregation(
+    model_type, num_communication_passes=5, tolerance=1e-6
+):
+  """ Create TFF function for robust aggregation.
+  The robust aggregate is an approximate geometric median
+  computed via the smoothed Weiszfeld algorithm.
+
+  Args:
+    model_type: tff typespec of quantity to be aggregated.
+    num_communication_passes: number of communication rounds in the
+      smoothed Weiszfeld algorithm (min. 1).
+    tolerance: smoothing parameter of smoothed Weiszfeld algorithm.
+      Default 1e-6.
+
+  Return:
+    An instance of `tff.utils.StatefulAggregateFn` which implements a
+    (stateless) robust aggregate.
+  """
+  if num_communication_passes < 1:
+    raise ValueError('Aggregation requires num_communication_passes >= 1')
+  # TODO: client weights have been hardcoded as float32
+
+  @tff.tf_computation(tf.float32, model_type, model_type)
+  def update_weight_fn(weight, server_model, client_model):
+    sqnorms = tf.nest.map_structure(lambda a, b: tf.norm(a-b)**2, server_model, client_model)
+    sqnorm = sum(six.itervalues(sqnorms))
```

Comment:
```suggestion
    sqnorm = tf.reduce_sum(list(six.itervalues(sqnorms)))
```

I'm not entirely sure if this makes a difference, but it may make sense to use the TF sum rather than the Python sum?